components/base-rest-api/packages/services/lib/token-swapper-service.js

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import _ from 'lodash';
+import { Service } from '@aws-ee/base-services-container';
+import jwtDecode from 'jwt-decode';
+import { newInvoker } from './authentication-providers/helpers/invoker';
+
+const settingKeys = {
+  dbValidTokens: 'dbValidTokens',
+};
+class TokenSwapperService extends Service {
+  constructor() {
+    super();
+    this.boom.extend(['invalidToken', 403]);
+    this.dependency(['dbService', 'authenticationProviderConfigService']);
+  }
+
+  async init() {
+    this.invoke = newInvoker(this.container.find.bind(this.container));
+  }
+
+  async swap({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    if (!uid) {
+      throw this.boom.badRequest('Empty uid', true);
+    }
+    let payload;
+    try {
+      payload = jwtDecode(token);
+    } catch (error) {
+      throw this.boom.invalidToken('Invalid Token', true).cause(error);
+    }
+    const ttl = _.get(payload, 'exp', 0);
+    if (ttl === 0) {
+      throw this.boom.invalidToken('Expired Token', true);
+    }
+    const { isSameToken } = await this.revokeValidToken({ token, uid });
+    if (isSameToken) {
+      return;
+    }
+    const record = { uid, token, ttl };
+    await dbService.helper
+      .updater()
+      .table(dbValidTokens)
+      .key({ uid })
+      .item(record)
+      .update();
+  }
+
+  async revokeValidToken({ token, uid }) {
+    const dbService = await this.service('dbService');
+    const authenticationProviderConfigService = await this.service('authenticationProviderConfigService');
+    const dbValidTokens = this.settings.get(settingKeys.dbValidTokens);
+
+    const record = await dbService.helper
+      .getter()
+      .table(dbValidTokens)
+      .key({ uid })
+      .get();
+
+    if (!record) {
+      return { isSameToken: false };
+    }
+
+    const oldToken = record.token;
+    const isSameToken = token === oldToken;
+    if (isSameToken) {
+      return { isSameToken: true };
+    }
+
+    let payload;
```

Comment:
You should have a unit test for the case where there is an invalid token in the db.